DIG-1671: Adding date format to manifest

yavyx opened the pull request:
### What's new
- Added a required date_format field to the manifest. It specifies the order of day, month and year to parse ambiguous dates properly.
- Updated test data and manifest with date formats.
- Improved mapping exceptions to be more readable and specific

### How to test
- Add date_format to the manifest
- Run CSVConvert.py using a dataset with ambiguous dates for that field and try different patterns in the template to confirm they are parsed as indicated.
- Test removing date resolutions and references to see the exceptions.

Previously:
~~Added a `date_format` parameter to the `date_interval` mapping function to specify the order of day, month and year to parse ambiguous dates properly.~~

~~How to test~~
~~- Add `date_format` to any date field in the template~~
~~DONOR.INDEX.date_of_birth, {date_interval(Donor.date_of_birth, date_format="DMY")}~~
~~- Run `CSVConvert.py` using a dataset with ambiguous dates for that field and try different patterns in the template to confirm they are parsed as indicated.~~

yavyx committed:
adding date format to date intervals

daisieh reviewed (changes requested):
If I change the template line to
```
DONOR.INDEX.date_of_birth, {date_interval(Donor.date_of_birth, date_format="MDY")}
```
I don't think the date_format gets passed in? If I add a print statement to the top of `date_interval` in mappings.py, like
```
def date_interval(data_values, date_format=None):
    print(f"hello {date_format}")
```
I get a lot of "hello None" in the output.

I don't think that you can pass in extra arguments like this into the template file, at least not as I've written it.

I appreciate the idea behind making it more flexible for date formats, but I'm not sure that it's possible to do it this way. I think you might have to write a wrapper method in your own .py file that parses the date with a format.

yavyx committed:
addded date format

yavyx committed:
updated docs

yavyx committed:
updated test date formats

yavyx committed:
make exceptions more readable

yavyx committed:
make exception specific

yavyx committed:
cleanup

yavyx committed:
cleanup

yavyx commented:
Re-submitting PR, [DIG-1671](https://candig.atlassian.net/browse/DIG-1671)

### What's new
- Added a required `date_format` field to the manifest. It specifies the order of day, month and year to parse ambiguous dates properly.
- Updated test data and manifest with date formats.
- Improved mapping exceptions to be more readable and specific

### How to test
- Add `date_format` to the manifest
- Run `CSVConvert.py` using a dataset with ambiguous dates for that field and try different patterns in the template to confirm they are parsed as indicated. 
- Test removing date resolutions and references to see the exceptions.

[DIG-1671]: https://candig.atlassian.net/browse/DIG-1671?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mshadbolt commented on src/clinical_etl/CSVConvert.py:
```diff
@@ -643,6 +646,10 @@ def csv_convert(input_path, manifest_file, minify=False, index_output=False, ver
         sys.exit("Need to specify what the main identifier column name is as 'identifier' in the manifest file, "
                  "see README for more details.")
+    mappings.DATE_FORMAT = manifest["date_format"]
+    if mappings.IDENTIFIER_FIELD is None:
```
Should this be mappings.DATE_FORMAT?
```suggestion
    if mappings.DATE_FORMAT is None:
```

yavyx replied:
Yes, thank you

yavyx committed:
Update src/clinical_etl/CSVConvert.py

Co-authored-by: Marion <mshadbolt@users.noreply.github.com>

yavyx committed:
exception when parser fails

yavyx committed:
update reference date docs

mshadbolt committed:
improved error catching in manifest

mshadbolt commented:
I added some try catch blocks to the part that is reading the manifest so that if the fields are blank, the script exits as well as if the manifest doesn't have those specified at all, also added a catch to ensure the date format is some form of 'DMY/MDY/YDM'

mshadbolt committed:
better error msg

mshadbolt reviewed (approved):
Looks good to me and works as expected

daisieh commented on src/clinical_etl/mappings.py:
```diff
@@ -79,22 +91,31 @@ def date_interval(data_values):
+    except KeyError:
+        raise MappingError("No reference date found to calculate date_interval: is there a reference_date specified in the manifest?", field_level=1)
+    DEFAULT_DATE_PARSER = dateparser.DateDataParser(
+        settings={"PREFER_DAY_OF_MONTH": "first", "DATE_ORDER": DATE_FORMAT}
```
It might be good to declare DATE_FORMAT at the top of this file, like we do with INDEXED_DATA or IDENTIFIER.

daisieh commented on src/clinical_etl/mappings.py:
```diff
@@ -79,22 +91,31 @@ def date_interval(data_values):
 
     Args:
         data_values: a values dict with a date
+        date_format: string with order of day, month and year in date (e.g. "DMY", "YMD", etc.)
```
This isn't really an argument anymore...should this be here, or is it described somewhere else?

daisieh reviewed (changes requested):
Tiny changes

mshadbolt committed:
fix month intervals

mshadbolt committed:
remove import

yavyx committed:
minor fixes

yavyx commented:
@daisieh I addressed the changes and Marion's fix is working

daisieh reviewed (approved):
Looks good to me! Thanks to both @yavyx and @mshadbolt for taking over so much of the coding in this project!
